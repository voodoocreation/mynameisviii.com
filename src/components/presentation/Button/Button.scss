@import "../../../scss/variables", "../../../scss/mixins";

.Button.isStyled {
  @include panel-1();

  display: inline-block;
  position: relative;
  padding: $spacing-half $spacing;
  border: $text-color 1px solid;
  border-radius: $radius;

  &.isLoading .Button--content {
    opacity: 0;
  }

  .Loader {
    display: flex;
    align-items: center;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;

    svg {
      margin: 0 auto;
    }
  }

  &:hover {
    @include panel-2();
  }

  @media print {
    color: #000;
    border-color: #000;
  }
}
