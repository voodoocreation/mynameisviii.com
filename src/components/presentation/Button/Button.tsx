import cn from "classnames";
import * as React from "react";

import Loader from "../Loader/Loader";

import "./Button.scss";

interface IProps {
  className?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  isStyled?: boolean;
  title?: string;
  type?: "button" | "submit";
  onClick: () => void;
  nodeRef?: React.RefObject<HTMLButtonElement>;
}

export default class Button extends React.Component<IProps> {
  public static defaultProps = {
    isDisabled: false,
    isLoading: false,
    isStyled: true,
    type: "button"
  };

  public render() {
    const {
      children,
      className,
      isDisabled,
      isLoading,
      isStyled,
      onClick,
      title,
      type
    } = this.props;

    return (
      <button
        className={cn("Button", className, { isLoading }, { isStyled })}
        disabled={isDisabled}
        onClick={!isLoading ? onClick : undefined}
        ref={this.props.nodeRef}
        title={title}
        type={type}
      >
        <span className="Button--content">{children}</span>
        {isLoading ? <Loader /> : null}
      </button>
    );
  }
}
