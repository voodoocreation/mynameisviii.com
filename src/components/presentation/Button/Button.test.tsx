import ComponentTester from "../../../utilities/ComponentTester";
import Button from "./Button";

const component = new ComponentTester(Button)
  .withDefaultProps({
    onClick: jest.fn()
  })
  .withDefaultChildren("Button text");

describe("[presentation] <Button />", () => {
  describe("when rendering with default props", () => {
    const { props, wrapper } = component.mount();

    beforeAll(() => {
      jest.clearAllMocks();
    });

    it("clicks the button", () => {
      wrapper.simulate("click");
    });

    it("calls the onClick prop", () => {
      expect(props.onClick).toHaveBeenCalledTimes(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when isLoading is true", () => {
    const { props, wrapper } = component
      .withProps({
        isLoading: true
      })
      .mount();

    beforeAll(() => {
      jest.clearAllMocks();
    });

    it("renders with isLoading class", () => {
      expect(wrapper.find(".isLoading")).toHaveLength(1);
    });

    it("renders with a Loader", () => {
      expect(wrapper.find("Loader")).toHaveLength(1);
    });

    it("clicks the button", () => {
      wrapper.simulate("click");
    });

    it("doesn't call the onClick prop", () => {
      expect(props.onClick).toHaveBeenCalledTimes(0);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });
});
