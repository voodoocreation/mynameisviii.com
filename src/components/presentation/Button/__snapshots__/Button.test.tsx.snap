// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Button /> when isLoading is true matches snapshot 1`] = `
<button
  class="Button isLoading isStyled"
  type="button"
>
  <span
    class="Button--content"
  >
    Button text
  </span>
  <span
    class="Loader"
  >
    <div
      class="Loader--animation"
    >
      <span />
      <span />
      <span />
    </div>
  </span>
</button>
`;

exports[`[presentation] <Button /> when rendering with default props matches snapshot 1`] = `
<button
  class="Button isStyled"
  type="button"
>
  <span
    class="Button--content"
  >
    Button text
  </span>
</button>
`;
